### PR TITLE
Valid if ANY of the range-specifier values overlap

### PR DESCRIPTION
**Pull request by jcready**
Fixes https://github.com/jshttp/range-parser/issues/13

### CODE REVIEW

**Comment by dougwilson** [comment]
Thanks! Please add tests and update the documentation.

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "range-parser",
   "author": "TJ Holowaychuk <tj@vision-media.ca> (http://tjholowaychuk.com)",
   "description": "Range header field string parser",
-  "version": "1.0.3",
+  "version": "1.1.0",
```
Comment (dougwilson) [inline package.json]: Please undo this change; bumping the version is part of the release process.

File: HISTORY.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+1.1.0 / 2016-03-21
```
Comment (dougwilson) [inline HISTORY.md]: Please undo adding a version on this line, as this is part of the release process.

**Comment by jcready** [comment]
Is there anything else you need?

**Comment by dougwilson** [comment]
Hi @jcready, sorry, I didn't realize you were ready for me to take another look :) Just waiting for you to update the documentation now.

**Comment by jcready** [comment]
Yeah, that's the part I'm not sure about. As far as I can tell the documentation is still accurate as it stands, only more so now since it's following the spec's definition of an invalid range.

**Comment by dougwilson** [comment]
I think it needs to call out that invalid ranges will be silently ignored and not included in the result (or any indication that any invalid range was included). Currently the interface preserves all ranges in the exact order they were in the header. This PR changes that behavior and it should be documented how this works.

**Comment by dougwilson** [comment]
At least describing it somewhere, like the changelog, as a change in behavior from how it operated previously.

**Comment by jcready** [comment]
@dougwilson I attempted to do that [here](https://github.com/jshttp/range-parser/pull/14/commits/c26fcb52546026bd0a3ac7aa10096306f7c7bb54#diff-88dc7475eedf918122374be6d7c2c151R4). I'll try to fit it in the README.md somehwere.

**Comment by dougwilson** [comment]
> @dougwilson I attempted to do that here. I'll try to fit it in the README.md somehwere.

Sure, I remember. Why did you remove it, then?

**Comment by jcready** [comment]
@dougwilson lol, because I'm not sure how to fit it into there without also adjusting the version number.

**Comment by jcready** [comment]
I mean, I guess I can and you can just add the version above it. Is that what you'd like?

**Comment by dougwilson** [comment]
Gotcha, I didn't know that :) If you asked, I would have given a pointer. Just take a look at any commit in the commit history, like https://github.com/jshttp/range-parser/commit/a00cb33936797a9321c49b8040a00dbd949d88e9

**Comment by jcready** [comment]
Unrelated, but have you considered including a function to combine the ranges?

```js
function combineRanges(ranges) {
  return ranges.sort(function (a, b){
    return a.start - b.start;
  }).reduce(function (memo, nextRange) {
    var currentRange = memo[memo.length - 1];
    if (!memo.length || nextRange.start - currentRange.end > 1) {
      memo.push(nextRange);
    } else if (nextRange.end > currentRange.end) {
      currentRange.end = nextRange.end;
    }
    return memo;
  }, [])
}
```

Example:

```js
combineRanges(rangeParser(150, 'bytes=0-4,90-99,5-75,100-199,101-102'))
> [{ "start": 0, "end": 75 }, { "start": 90, "end": 149 }]
```

**Comment by jcready** [comment]
Is there anything else you need from me to merge this?

**Comment by dougwilson** [comment]
Hi @jcready, I'm really sorry this issue fell through the cracks :( I'm merging it now! As for the idea in https://github.com/jshttp/range-parser/pull/14#issuecomment-199959982 to combine ranges, that is a great idea and would happily accept a PR if you want to make it. I promise I won't let it sit like this one did :(

**Comment by jcready** [comment]
@dougwilson how would you prefer I include the combine ranges thing?

1. By simply inlining the functionality into the original function?

   ```js
   rangeParser(150, 'bytes=0-4,90-99,5-75,100-199,101-102')
   > [{ "start": 0, "end": 75 }, { "start": 90, "end": 149 }]
   ```

2. As a property on the `rangeParser` function that takes the result of calling `rangeParser` as input?

   ```js
   rangeParser.combineRanges(rangeParser(150, 'bytes=0-4,90-99,5-75,100-199,101-102'))
   > [{ "start": 0, "end": 75 }, { "start": 90, "end": 149 }]
   ```

3. Or as a property on the `rangeParser` function that takes `size, str` as input and calls `rangeParser` itself?

   ```js
   rangeParser.combineRanges(150, 'bytes=0-4,90-99,5-75,100-199,101-102')
   > [{ "start": 0, "end": 75 }, { "start": 90, "end": 149 }]
   ```

**Comment by dougwilson** [comment]
Hi @jcready, I was thinking abut this and I think based on how the current interface to this module is, probably the best fit is to just alter the existing behavior (i.e. option 1). This has the benefit of being safer out-of-the-box such that clients can't easily just ask for the same pieces over and over again in the same request. If there ever pops up a use-case for needing to access the un-combined ranges, we can always add that in as a secondary, raw API.
